ops/testing.py

```diff
@@ -1141,6 +1141,67 @@ def reset_planned_units(self):
         """
         self._backend._planned_units = None
 
+    def add_network(self, binding_name: str, address: str, *,
+                    relation_id: Optional[int] = None,
+                    cidr: Optional[str] = None,
+                    interface: str = 'eth0',
+                    ingress_addresses: Optional[List[str]] = None,
+                    egress_subnets: Optional[List[str]] = None,
+                    ):
+        """Add simulated network data for the given binding.
+
+        Calling this multiple times with the same (binding_name, relation_id)
+        combination will replace the associated network data.
+
+        This method only supports the common case of adding a single interface
+        and bind address. If you need support for multiple bind addresses, you
+        can patch ops.testing._TestingModelBackend.network_get and return the
+        full network-get data via the patched method.
+
+        Args:
+            binding_name: Name of binding (endpoint) to add network data for.
+                Use "" (empty string) to set the default binding.
+            address: Binding's IP address, for example "10.0.0.10".
+            relation_id: Optional relation ID for the binding. If not
+                provided, adds network data for the default binding.
+            cidr: Binding's CIDR. Defaults to "<address>/24".
+            interface: Name of network interface.
+            ingress_addresses: Optional list of ingress addresses. Defaults to
+                [address].
+            egress_subnets: Optional list of egress subnets. Defaults to
+                [cidr]
+        """
```

Comment:
Can this method raise any exceptions?

Reply:
Not really any interesting ones, no. (Just standard ones like TypeError if you pass in the wrong type for `binding_name` or something silly. But we're starting to lean in on type annotations for that anyway.)

Reply:
That said, I just updated it to check that `address` was an IPv4 or IPv6 address and noted that ValueError is raised if not.